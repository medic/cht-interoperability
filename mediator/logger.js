const {format, createLogger, transports} = require('winston');
const isProduction = process.env.NODE_ENV === 'production';

const logger = createLogger({
  level: !isProduction ? 'debug' : 'info',
  format: format.combine(
    format.splat(),
    format.simple()
  ),
  transports: [
    new transports.Console({
      format: format.combine(
        format(info => {
          info.level = info.level.toUpperCase();
          return info;
        })(),
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(
          info =>
          `${info.timestamp} ${info.level}: ${info.message} ${info.stack ? info.stack : ''}`
          )
      ),
    }),
  ],
});

module.exports = logger;
