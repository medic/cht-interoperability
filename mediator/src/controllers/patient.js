const axios = require('axios');
const { genereateFHIRPatientResource } = require('../utils/patient');
const FHIR_ROOT_URL = 'http://openhim-core:5001/fhir';

async function createPatient(CHTpatientDoc) {
  const FHITPatientResource = genereateFHIRPatientResource(CHTpatientDoc);

  try {
    const res = await axios.post(`${FHIR_ROOT_URL}/Patient`, FHITPatientResource, { auth: {
      username: 'interop-client',
      password: 'interop-password'
    }});
    return {status: res.status, patient: res.data}
  } catch (error) {
    console.error(error);
    return {status: error.status, patient: error.data}
  }
}

module.exports = {
  createPatient
};
