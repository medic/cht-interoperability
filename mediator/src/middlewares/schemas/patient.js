const joi = require('joi');

const createPatientSchema = joi.object({
  name: joi.string().optional(),
  _id: joi.string().optional(),
  sex: joi.string().optional(),
  birthDate: joi.string().optional(),
});

module.exports = {
  createPatientSchema,
};
