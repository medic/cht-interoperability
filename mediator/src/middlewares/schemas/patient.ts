import joi from "joi";


export const createPatientSchema = joi.object({
  parent: joi.any().optional(),
  type: joi.string().optional(),
  name: joi.string().required(),
  _id: joi.string().required(),
  id: joi.string().optional(),
  sex: joi.string().trim().valid('male', 'female', 'other', 'unkown').required(),
  // validate expecting YYYY-MM-DD
  date_of_birth: joi
    .string()
    .regex(/((?:19|20)\d\d)-(0?[1-9]|1[012])-([12][0-9]|3[01]|0?[1-9])/)
    .required()
    .messages({
      "object.regex": "Invalid date expecting YYYY-MM-DD",
      "string.pattern.base": "Invalid date expecting YYYY-MM-DD"
    })
});
