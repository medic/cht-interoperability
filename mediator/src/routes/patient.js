const {Router} = require('express');
const router = Router();
const patientController = require('../controllers/patient');
const {validateBodyAgainst} = require('../middlewares');
const {createPatientSchema} = require('../middlewares/schemas/patient');

router.post('/',
  validateBodyAgainst(createPatientSchema),
  async function(req, res) {
    const {status, patient} = await patientController.createPatient(req.body);
    res.send({status, patient});
});

module.exports = router;
