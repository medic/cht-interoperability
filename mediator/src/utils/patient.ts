function genereateFHIRPatientResource(patient: any) {
  const patientLastName = patient.name.split(' ').slice(-1);
  const FHITPatientResource = {
    resourceType: 'Patient',
    id: patient.id,
    identifier: [
      {
        system: 'cht',
        value: patient._id
      }
    ],
    name: [
      {
        use: 'official',
        family: patientLastName,
        given: [patient.name]
      }
    ],
    gender: patient.sex,
    birthDate: patient.date_of_birth
  };

  return FHITPatientResource;
}

module.exports = {
  genereateFHIRPatientResource,
};
