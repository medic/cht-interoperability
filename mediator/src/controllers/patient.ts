import axios from "axios";
import { FHIR } from "../../config";
import { logger } from "../../logger";

const { url: fhirUrl, username: fhirUsername, password: fhirPassword } = FHIR;

export async function createPatient(patient: fhir5.Patient) {
  try {
    const res = await axios.post(`${fhirUrl}/Patient`, patient, {
      auth: {
        username: fhirUsername,
        password: fhirPassword,
      },
    });
    return { status: res.status, data: res.data };
  } catch (error: any) {
    logger.error(error);
    return { status: error.status, data: error.data };
  }
}
