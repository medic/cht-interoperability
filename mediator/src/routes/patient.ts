import { Router } from 'express';
import { validateBodyAgainst } from '../middlewares';
import { createFhirResource, validateFhirResource } from '../utils/fhir';
import { PatientSchema } from '../middlewares/schemas/patient';
import { requestHandler } from '../utils/request';

const router = Router();

const resourceType = 'Patient';

router.post(
  '/',
  validateBodyAgainst(validateFhirResource(resourceType), PatientSchema),
  requestHandler((req) => createFhirResource({ ...req.body, resourceType }))
);

export default router;
