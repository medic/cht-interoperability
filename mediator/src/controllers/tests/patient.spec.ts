import axios from "axios";
import { logger } from "../../../logger";
import { createPatient } from "../patient";

jest.mock("../../../logger");
jest.mock("axios");
const mockAxios = axios as jest.Mocked<typeof axios>;

describe("Patient controllers", () => {
  const patient: fhir5.Patient = {
    resourceType: "Patient"
  };

  describe("createPatient", () => {
    it("creates a patient when given a valid patient resource", async () => {
      const data = { status: 200, data: {} };

      mockAxios.post.mockResolvedValueOnce(data);

      const res = await createPatient(patient);

      expect(res.data).toBe(data.data);
      expect(res.status).toEqual(data.status);
      expect(mockAxios.post).toHaveBeenCalled();
      expect(mockAxios.post).toMatchSnapshot();
      expect(logger.error).not.toHaveBeenCalled();
    });

    it("handles error when the patient creation fails", async () => {
      const data = { status: 400, data: {} };

      mockAxios.post.mockRejectedValueOnce(data);

      const res = await createPatient(patient);

      expect(res.status).toBe(data.status);
      expect(logger.error).toHaveBeenCalled();
      expect(mockAxios.post).toHaveBeenCalled();
      expect(res.data).toBe(res.data);
    });
  });
});
