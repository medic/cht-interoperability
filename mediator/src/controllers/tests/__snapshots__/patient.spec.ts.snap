// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Patient controllers createPatient creates a patient when given a valid patient resource 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Patient",
      {
        "resourceType": "Patient",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
